? .README.swp
? .tmp_versions
Index: at76c503.c
===================================================================
RCS file: /cvsroot/at76c503a/at76c503a/at76c503.c,v
retrieving revision 1.71
diff -u -r1.71 at76c503.c
--- at76c503.c	14 Oct 2004 01:21:21 -0000	1.71
+++ at76c503.c	15 Oct 2004 03:14:52 -0000
@@ -3912,7 +3912,7 @@
 	}
 
 	eth_hdr = (struct ethhdr *)(skb->data-sizeof(struct ethhdr));
-	skb->mac.ethernet = eth_hdr;
+	skb->mac.raw = eth_hdr;
 	if (build_ethhdr) {
 		/* This needs to be done in this order (eth_hdr->h_dest may
 		 * overlap src_addr) */
